src/main/java/org/elasticsearch/common/util/MultiDataPathUpgrader.java

```diff
@@ -0,0 +1,366 @@
+             CheckIndex checkIndex = new CheckIndex(directory)) {
+            checkIndex.setInfoStream(out);
+            CheckIndex.Status status = checkIndex.checkIndex();
+            out.flush();
+            if (!status.clean) {
+                logger.warn("check index [failure]\n{}", new String(os.bytes().toBytes(), Charsets.UTF_8));
+                throw new ElasticsearchIllegalStateException("index check failure");
+            }
+        }
+    }
+
+    /**
+     * Returns true iff the given shard needs upgrading.
+     */
+    public boolean needsUpgrading(ShardId shard) {
+        final Path[] paths = nodeEnvironment.shardPaths(shard);
+         // custom data path doesn't need upgrading neither single path envs
+        if (paths.length > 1) {
+            int numPathsExist = 0;
+            for (Path path : paths) {
+                if (Files.exists(path.resolve(MetaDataStateFormat.STATE_DIR_NAME))) {
+                    numPathsExist++;
+                    if (numPathsExist > 1) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Picks a target ShardPath to allocate and upgrade the given shard to. It picks the target based on a simple
+     * heuristic:
+     * <ul>
+     *  <li>if the smallest datapath has 2x more space available that the shards total size the datapath with the most bytes for that shard is picked to minimize the amount of bytes to copy</li>
+     *  <li>otherwise the largest available datapath is used as the target no matter how big of a slice of the shard it already holds.</li>
+     * </ul>
+     */
+    public ShardPath pickShardPath(ShardId shard) throws IOException {
+        if (needsUpgrading(shard) == false) {
+            throw new IllegalStateException("Shard doesn't need upgrading");
+        }
+        final NodeEnvironment.NodePath[] paths = nodeEnvironment.nodePaths();
+
+        // if we need upgrading make sure we have all paths.
+        for (NodeEnvironment.NodePath path : paths) {
+            Files.createDirectories(path.resolve(shard));
+        }
+        final ShardFileInfo[] shardFileInfo = getShardFileInfo(shard, paths);
+        long totalBytesUsedByShard = 0;
+        long leastUsableSpace = Long.MAX_VALUE;
+        long mostUsableSpace = Long.MIN_VALUE;
+        assert shardFileInfo.length ==  nodeEnvironment.shardPaths(shard).length;
+        for (ShardFileInfo info : shardFileInfo) {
+            totalBytesUsedByShard += info.spaceUsedByShard;
+            leastUsableSpace = Math.min(leastUsableSpace, info.usableSpace + info.spaceUsedByShard);
+            mostUsableSpace = Math.max(mostUsableSpace, info.usableSpace + info.spaceUsedByShard);
+        }
+
+        if (mostUsableSpace < totalBytesUsedByShard) {
+            throw new IllegalStateException("Can't upgrade path available space: " + mostUsableSpace + " bytes required space: " + totalBytesUsedByShard + " bytes");
```

Comment:
can we convert mostUsableSpace and totalBytesUsedByShard to ByteSizeValue when reporting? this will make it user readable (at the expense of precision, which I think is fine?)